stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -404,21 +382,42 @@ static bool swift_task_hasTaskGroupStatusRecordImpl() {
 /**************************************************************************/
 
 // ==== Child tasks ------------------------------------------------------------
-SWIFT_CC(swift)
-static ChildTaskStatusRecord*
-swift_task_attachChildImpl(AsyncTask *child) {
-  void *allocation = malloc(sizeof(swift::ChildTaskStatusRecord));
-  auto record = new (allocation) swift::ChildTaskStatusRecord(child);
-  SWIFT_TASK_DEBUG_LOG("attach child task = %p, record = %p, to current task = %p",
-                       child, record, swift_task_getCurrent());
-  swift_task_addStatusRecord(record);
-  return record;
-}
 
+/* Called in the path of linking a child into a parent/group synchronously with
+ * the parent task.
+ *
+ * When called to link a child into a parent directly, this does not hold the
+ * parent's task status record lock. When called to link a child into a task
+ * group, this holds the parent's task status record lock.
+ */
+SWIFT_EXPORT_FROM(swift_Concurrency)
 SWIFT_CC(swift)
-static void
-swift_task_detachChildImpl(ChildTaskStatusRecord *record) {
-  swift_task_removeStatusRecord(record);
+void swift_task_updateNewChildWithParentAndGroupState(
+    AsyncTask *child, ActiveTaskStatus parentStatus, TaskGroup *group) {
+  /*
+   * We can take the fast path of just modifying the ActiveTaskStatus in the
+   * child task since we know that it won't have any task status records and
+   * cannot be accessed by anyone else since it hasn't been linked in yet.
+   * Avoids the extra logic in `swift_task_cancel` and `swift_task_escalate`
+   */
+  auto oldChildTaskStatus =
+      child->_private().Status.load(std::memory_order_relaxed);
+  assert(oldChildTaskStatus.getInnermostRecord() == NULL);
+
+  auto newChildTaskStatus = oldChildTaskStatus;
+
+  /* Parent task is cancelled or group the child task is part of (if any) is
+   * cancelled */
+  if (parentStatus.isCancelled() || (group && group->isCancelled())) {
+    newChildTaskStatus = newChildTaskStatus.withCancelled();
+  }
+
+  /* Parent task got escalated, make sure to propagate it to child. */
+  if (parentStatus.isStoredPriorityEscalated()) {
+    newChildTaskStatus = newChildTaskStatus.withEscalatedPriority(
+        parentStatus.getStoredPriority());
+  }
```

Comment:
I'm not sure it's safe to rely on `isStoredPriorityEscalated` here, since it can be cleared asynchronously.  You should just set the priority of the child task to the priority of the parent.  You should also update the priority in the Job header.

Reply:
How can it be cleared? Isn't escalation permanent on a task?

I'd expect a task to be created at say, UT priority, and then if it gets escalated to IN priority, to have the escalated flag be permanently set on the task.

Reply:
Oh okay reading further, I think I misunderstood the semantics of the `isEscalated` flag. Okay you're right, I need to change this.